### PR TITLE
TT-968 Add k8schaos library to use Chaos Mesh as code

### PR DESCRIPTION
**Pull request by lukaszcl**
The `k8schaos` package is a Go library designed to facilitate chaos testing within Kubernetes environments using Chaos Mesh. It offers a structured way to define, execute, and manage chaos experiments as code, directly integrated into Go applications or testing suites. This package simplifies the creation and control of Chaos Mesh experiments, including network chaos, pod failures, and stress testing on Kubernetes clusters.

### Features
- **Chaos Object Management:** Easily create, update, pause, resume, and delete chaos experiments using Go structures and methods.
- **Lifecycle Hooks:** Utilize chaos listeners to hook into lifecycle events of chaos experiments, such as creation, start, pause, resume, and finish.
- **Support for Various Chaos Experiments:** Create and manage different types of chaos experiments like NetworkChaos, IOChaos, StressChaos, PodChaos, and HTTPChaos.
- **Chaos Experiment Status Monitoring:** Monitor and react to the status of chaos experiments programmatically.

### CODE REVIEW

File: k8schaos/logger.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// Default logger
+var Logger zerolog.Logger
+
+func init() {
+	// Default logger
+	Logger = CreateLogger(LoggerConfig{
```
Comment (skudasov) [inline k8schaos/logger.go:15]: Oh, we have a package-level logger, maybe add it to docs as the simplest example?

File: k8schaos/chaos_helper.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package k8schaos
+
+import (
+	"errors"
+	"time"
+)
+
+func WaitForAllChaosRunning(chaosObjects []*Chaos, timeoutDuration time.Duration) error {
```
Comment (skudasov) [inline k8schaos/chaos_helper.go:8]: I think we should add it to docs

File: k8schaos/chaos.go [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+	default:
+		return fmt.Errorf("unsupported chaos object type: %T", obj)
+	}
+
+	return nil
+}
+
+func isConditionTrue(status *v1alpha1.ChaosStatus, expectedCondition v1alpha1.ChaosCondition) bool {
+	if status == nil {
+		// c.loggerog.Info().Msg("experiment is missing status information")
```
Comment (skudasov) [inline k8schaos/chaos.go:413]: Let's uncomment it, if it makes sense. More logs are better

File: k8schaos/schedule.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			if err := s.Client.Get(ctx, client.ObjectKey{
+				Namespace: s.Object.GetNamespace(),
+				Name:      s.Object.GetName(),
+			}, &schedule); err != nil {
+				Logger.Error().Err(err).Msg("Failed to get Schedule object")
+				continue
+			}
+
+			// Log or process the schedule's current status
+			// Loggerog.Info().Interface("status", schedule.Status).Msg("Current Schedule Status")
```
Comment (skudasov) [inline k8schaos/schedule.go:182]: Let's uncomment it, if it makes sense. More logs are better

File: k8schaos/schedule.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+type ScheduleStatus string
+
+const (
+	ScheduleStatusCreated ScheduleStatus = "created"
+	ScheduleStatusDeleted ScheduleStatus = "deleted"
+	ScheduleStatusUnknown ScheduleStatus = "unknown" // For any state that doesn't match the above
+)
+
+type Schedule struct {
```
Comment (skudasov) [inline k8schaos/schedule.go:20]: I'd like to hear the use case. If we will define all our experiments as code why do we need a Schedule then? If we go this way we can just define the whole set of experiments as one schedule in YAML. Do you think it's useful to combine experiments as a code and their internal logic to do a schedule?

File: k8schaos/chaos.go [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+	K8sClient   client.Client
+}
+
+func (o *NetworkChaosOpts) Validate() error {
+	if o.Delay != nil {
+		latency, err := time.ParseDuration(o.Delay.Latency)
+		if err != nil {
+			return fmt.Errorf("invalid latency: %v", err)
+		}
+		if latency > 500*time.Millisecond {
```
Comment (skudasov) [inline k8schaos/chaos.go:569]: This validation is product specific, should we move it to the product repo? I propose that we expose only basic API and users can do their validation
I'm not sure we'd like to support these wrappers, they are pretty easy to create and there is no need to keep them centralized or, at least, let's make it less product specific, I may require to have short 1-2s delay in my cases

File: k8schaos/chaos.go [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+	c.notifyListeners(string(StatusCreated), nil)
+
+	// Create a cancellable context for monitorStatus
+	monitorCtx, cancel := context.WithCancel(ctx)
+	c.cancelMonitor = cancel
+	go c.monitorStatus(monitorCtx)
+}
+
+func (c *Chaos) Pause(ctx context.Context) error {
+	c.updateChaosObject(ctx)
```
Comment (skudasov) [inline k8schaos/chaos.go:134]: Please, handle the err

File: k8schaos/chaos.go [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+
+func (c *Chaos) Delete(ctx context.Context) error {
+	// Cancel the monitoring goroutine
+	if c.cancelMonitor != nil {
+		c.cancelMonitor()
+	}
+
+	// If the chaos was running or paused, update the status and notify listeners
+	if c.Status == StatusPaused || c.Status == StatusRunning {
+		c.updateChaosObject(ctx)
```
Comment (skudasov) [inline k8schaos/chaos.go:166]: Please, handle the err

File: k8schaos/chaos.go [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+	events := &corev1.EventList{}
+	if err := c.Client.List(context.Background(), events, listOpts...); err != nil {
+		return nil, fmt.Errorf("could not list chaos events: %w", err)
+	}
+
+	return events, nil
+}
+
+func (c *Chaos) GetChaosKind() string {
+	_, ok := c.Object.(*v1alpha1.NetworkChaos)
```
Comment (skudasov) [inline k8schaos/chaos.go:268]: Should we use a `(type)` switch for consistency?

File: k8schaos/chaos.go [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+	_, ok = c.Object.(*v1alpha1.HTTPChaos)
+	if ok {
+		return "HTTPChaos"
+	}
+
+	panic(fmt.Sprintf("could not get chaos kind for object: %v", c.Object))
+}
+
+func (c *Chaos) GetChaosStatus() (*v1alpha1.ChaosStatus, error) {
+	net, ok := c.Object.(*v1alpha1.NetworkChaos)
```
Comment (skudasov) [inline k8schaos/chaos.go:293]: Should we use a `(type)` switch for consistency?

File: k8schaos/chaos.go [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+	}
+	httpChaos, ok := c.Object.(*v1alpha1.HTTPChaos)
+	if ok {
+		return httpChaos.GetStatus(), nil
+	}
+	return nil, fmt.Errorf("could not get chaos status for %s", c.GetChaosKind())
+}
+
+func (c *Chaos) GetExperimentStatus() (v1alpha1.ExperimentStatus, error) {
+	networkChaos, ok := c.Object.(*v1alpha1.NetworkChaos)
```
Comment (skudasov) [inline k8schaos/chaos.go:317]: Should we use a `(type)` switch for consistency?

File: k8schaos/chaos.go [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+				case StatusRunning:
+					c.startTime = time.Now()
+					c.notifyListeners("started", nil)
+				case StatusPaused:
+					c.notifyListeners("paused", nil)
+				case StatusFinished:
+					c.endTime = time.Now()
+					c.notifyListeners("finished", nil)
+					// Delete the chaos object when it finishes
+					c.Delete(context.Background())
```
Comment (skudasov) [inline k8schaos/chaos.go:544]: Please, handle the err

File: k8schaos/schedule.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+	done := make(chan struct{})
+	deleteTimer := time.NewTimer(s.Duration)
+	go func() {
+		select {
+		case <-ctx.Done():
+			// Context was canceled, ensure chaos object is deleted
+			if !deleteTimer.Stop() {
+				<-deleteTimer.C // Drain the timer if it already fired
+			}
+			s.Delete(context.Background())
```
Comment (skudasov) [inline k8schaos/schedule.go:140]: Please, handle the err

File: k8schaos/schedule.go [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+			close(done)
+		case <-deleteTimer.C:
+			// Duration elapsed, delete the chaos object
+			s.Delete(context.Background())
```
Comment (skudasov) [inline k8schaos/schedule.go:144]: Please, handle the err

**Review by skudasov — CHANGES_REQUESTED** [review]
LGTM, though, added some nitpicks

I tried to use it and I think it makes sense to expose only basic API so people can create their chaos suites (wrappers) and add their custom validation. I'm not sure we need to maintain experiment wrappers in this repo, since we are giving you full control with code it's okay to assemble experiments from `k8s` types as you see it.

I'm not sure which approach we pick in the end but I don't see a lot of use cases where I need to assert chaos results from code or have precise control over chaos experiments.
Also, it's a little bit easier for me just to use files and copy some examples from `ChaosMesh` docs or from generated stuff.

There are pros and cons of creating chaos experiments as code

PROS:
- Full control
- Types

CONS:
- Non-exhaustive, otherwise we need to scan the namespace and generate something
- K8s dependencies (I already faced it when integrating, will happen in the future)
- Creating code wrappers where you can just use plain files with static labels, can you give us more use cases where we need control from code to assert something?
- Many options, schedule or control experiments on our side and schedule them?

I haven't found any use case that requires full control, chaos testing should be black-box testing IMO we simulate realistic chaos that might happen and watch alerts. As for types, I can import ChaosMesh CRD definition and validate YAML files in my IDE pretty simply, close to typing without code wrappers.

Let's consolidate our approach, fix the nitpicks, and merge 🚀 .

File: k8schaos/chaos.go [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+	StatusUnknown        ChaosStatus = "unknown" // For any state that doesn't match the above
+)
+
+type ChaosOpts struct {
+	Object      client.Object
+	Description string
+	DelayCreate time.Duration
+	Client      client.Client
+	Listeners   []ChaosListener
+	Logger      *zerolog.Logger
```
[the remark this answers is not in the document]
Comment (lukaszcl) [reply k8schaos/chaos.go:51]: @skudasov there is only one logger for `Chaos` struct. And it's either default or custom if you provide it in `ChaosOpts`

File: k8schaos/console_logger.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+)
+
+type ChaosLogger struct {
+	logger zerolog.Logger
+}
+
+func NewChaosLogger() *ChaosLogger {
+	return &ChaosLogger{
+		logger: CreateLogger(LoggerConfig{
+			LogOutput: os.Getenv("CHAOS_LOG_OUTPUT"),
```
[the remark this answers is not in the document]
Comment (lukaszcl) [reply k8schaos/console_logger.go:17]: I will require it explicitly

File: k8schaos/chaos.go (hunk 11 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:317]: yes, I'll do it

File: k8schaos/chaos.go (hunk 10 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:293]: yes, I'll do it

File: k8schaos/chaos.go (hunk 12 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:544]: will do

File: k8schaos/schedule.go (hunk 13 above)
Comment (lukaszcl) [reply k8schaos/schedule.go:140]: will do

File: k8schaos/schedule.go (hunk 5 above)
Comment (lukaszcl) [reply k8schaos/schedule.go:21]: @skudasov `Schedule` is really useful. Let's say you want 2 nodes down in the DON for the whole time. When you just do `podchaos` then nodes may be recreated by the replica. `Schedule` makes sure that the chaos is constantly applied making sure that the 2 nodes are **always** down. I use it in Mercury tests https://github.com/smartcontractkit/mercury-pipeline/pull/754

File: k8schaos/chaos.go (hunk 6 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:556]: @skudasov this one I'd keep here. I saw a lot of issues with chaos mesh when latency was too high. Chaos was not applied and there were network related errors on the chaos controller server. I also could not remove this networkchaos object from Chaos Mesh. General validation should prevent this

File: k8schaos/chaos.go (hunk 7 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:133]: ok

File: k8schaos/chaos.go (hunk 8 above)
Comment (lukaszcl) [reply k8schaos/chaos.go:165]: ok

File: k8schaos/chaos.go (hunk 9 above)
Comment (lukaszcl) [reply k8schaos/chaos.go]: yes

File: k8schaos/console_logger.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+		Str("namespace", schedule.GetObject().GetNamespace()).
+		Str("description", schedule.GetChaosDescription()).
+		Str("duration", duration.String()).
+		Time("startTime", schedule.GetStartTime()).
+		Time("endTime", schedule.GetEndTime()).
+		Interface("spec", schedule.GetChaosSpec()).
+		Msg("Chaos schedule deleted")
+}
+
+func (l *ChaosLogger) commonChaosLog(logLevel string, chaos Chaos) *zerolog.Event {
```
[the remark this answers is not in the document]
Comment (lukaszcl) [reply k8schaos/console_logger.go]: fixed

**Review by skudasov — APPROVED** [review]

**Pull request merged**